[10.0][Add] sale_manual_delivery

leemannd opened the pull request:
Sale Manual Delivery
====================
The Goal of this module is to allow the manual creation of delivery slips. If activated,
a sale order won't direcly impact directly the stock. It will not make a reservation on the stock.
It allows the delivery and the impact on stock to be done manually when needed.
The goal is to be used on long term projects where not all the material is shipped at once.

leemannd committed:
base module & sales_team

leemannd committed:
Add base for wizard

leemannd committed:
wizard for manual procurements

yvaucher commented on sale_manual_delivery/wizard/line.py:
```diff
@@ -0,0 +1,35 @@
+from odoo import models, fields
+
+
+class ManualProcurement(models.TransientModel):
```
ManualLine

elicoidal commented on sale_manual_delivery/README.md:
```diff
@@ -0,0 +1,57 @@
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+Sale Manual Delivery
```
@leemannd   Thanks for your contribution to the OCA. 
Let me introduce to you some basic information about contribution in OCA and I will review other points in more comments inside the code.
In general, please follow the principles: 
* Contribution guidelines: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md
* Please follow the OCA README template:
https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
* We recommend to include a screenshot as well to help new comers (nice to have)
* Please use simple headers in all py files following OCA templates
https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

elicoidal replied:
in particular here: you need to add a line as document title

elicoidal commented on sale_manual_delivery/README.md:
```diff
@@ -0,0 +1,57 @@
+Sale Manual Delivery
+====================
+
+The Goal of this module is to allow the manual creation of delivery slips. If activated,
```
s/Goal/goal

elicoidal commented on sale_manual_delivery/README.md:
```diff
@@ -0,0 +1,57 @@
+The goal is to be used on long term projects where not all the material is shipped at once.
+
+Usage
+=====
```
remove the 2 empty sections

elicoidal commented on sale_manual_delivery/README.md:
```diff
@@ -0,0 +1,57 @@
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/sale-workflow/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
```
Update according newest template

elicoidal commented on sale_manual_delivery/README.md:
```diff
@@ -0,0 +1,57 @@
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+To contribute to this module, please visit http://odoo-community.org.
```
https

elicoidal commented on sale_manual_delivery/__manifest__.py:
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
```
please add short headers

elicoidal commented on sale_manual_delivery/__manifest__.py:
```diff
@@ -0,0 +1,26 @@
+    'category': 'Sale',
+    'author': 'Camptocamp SA, Odoo Community Association (OCA)',
+    'version': '10.0.1.0.0',
+    'sequence': 150,
```
is it necessary?

elicoidal commented on sale_manual_delivery/__manifest__.py:
```diff
@@ -0,0 +1,26 @@
+        'views/sale_order_view.xml',
+        'wizard/manual_proc_view.xml',
+    ],
+    'demo': [
```
remove empty key

elicoidal commented on sale_manual_delivery/models/procurement.py:
```diff
@@ -0,0 +1,13 @@
+class ProcurementOrder(models.Model):
+    _inherit = 'procurement.order'
+
+    carrier_id = fields.Many2one(
```
this looks irrelevant in the current context (module for manual picking, no carrier id)

leemannd replied:
@elicoidal Thanks for the feedbacks. One of the goals of this manual delivery, is to separate the delivery from the sale. The sale order is made once in the beginning of the project. Then parts are shipped later and be specified with another carrier.

elicoidal replied:
@leemannd Then maybe you can add it to the module README to understand it better?

elicoidal reviewed (changes requested):
Light review about format as the module is on WIP process

leemannd committed:
Update base computation

leemannd committed:
update to OCA guidelines

jcoux committed:
Add action_manual_procurement_wizard

jcoux committed:
Add action_manual_procurement_wizard in view

leemannd committed:
Merge pull request #1 from jcoux/patch-1

Add action_manual_procurement_wizard

leemannd committed:
Merge pull request #2 from jcoux/patch-2

Add action_manual_procurement_wizard in view

leemannd committed:
flake8 & OCA standards

jgrandguillaume committed:
[IMP] report manual procurement on SO from sale team

jgrandguillaume committed:
[ADD] Translation and .pot file

jgrandguillaume committed:
[TYP] Change manual proc for manual delivery

jgrandguillaume committed:
[IMP] Toggle manual,standard as a button in SO + UI

jgrandguillaume committed:
[IMP] Update translation files

jgrandguillaume committed:
[IMP] lot's of functionnal improvments

Don't add already planned or shipped lines in wizard
Ensure you have one picking per manual delivery
Forbid to add lines or ship more than ordered
Planned date is required
Improve views, naming and buttons
Better Readme
Default Carrier set from SO

jgrandguillaume committed:
[FIX] Do not add product of type 'service' in manual delivery creation

jgrandguillaume committed:
[ADD] Tests on various use cases

leemannd committed:
Merge pull request #3 from jgrandguillaume/jge-fix-improvms-sale-man-delivery

Jge fix improvements sale manual delivery

jgrandguillaume commented:
Hi,

I've functionally reviewed this module and made some improvements on the general usage, tests and flow. It has now a smooth way of dealing with manual delivery creation. It inform the user of the already created delivery and controls that he makes no mistakes (e.g. ship more than ordered).

On my side, I'm :+1: 

Regards,

Joël

jgrandguillaume reviewed (approved):
Functional review ok

mdietrichc2c committed:
Regroup picking per date_planned / carrier + correct wizard action from SO

vrenaville committed:
[FIX] picking split criteria

vrenaville committed:
[FIX] call super in case of standard delivery

vrenaville committed:
[FIX] when multiple picking candidate available take the right one with date corresponding

[FIX] revert of renaming field

JordiBForgeFlow commented:
@leemannd very nice and useful. Would be great to have it also in purchases. Only when the supplier confirms what is incoming will users create the appropiate incoming shipments , to match whst the supplier confirms as being shipped.

leemannd commented:
@elicoidal It has been updated

elicoidal commented on sale_manual_delivery/README.rst:
```diff
@@ -0,0 +1,59 @@
+=====================
+
+The goal of this module is to allow the manual creation of delivery slips. If activated,
+a sale order won't direcly impact directly the stock. It will not make a reservation on the stock.
```
directly is duplicated

mdietrichc2c committed:
Add product name to manual delivery lines

aheficent commented on sale_manual_delivery/README.rst:
```diff
@@ -0,0 +1,59 @@
+as a planned date. A wizard helps you to chose what to deliver by showing you how much you already 
+planned to ship.
+
+
```
I miss the usage section here

AaronHForgeFlow commented on sale_manual_delivery/wizard/manual_proc_view.xml:
```diff
@@ -0,0 +1,57 @@
+  </record>
+
+  <act_window id='action_wizard_manual_delivery'
+              name='Create Manual Delivery'
```
I'm not sure if this is good. I'm getting a singleton error when I select more than one SO in the list view and then I click on the action.
![image](https://cloud.githubusercontent.com/assets/19620251/25221921/16b14f6e-25b7-11e7-863d-26bc611ba0b3.png)

leemannd replied:
@aheficent Hello, Thank you for your feedback.
The wizard was intended to be launched from the SO and not from many. Currently I don't have the time to make improvements to this PR, but you can submit your idea. I'll be happy to review and include your commits into this PR.

AaronHForgeFlow replied:
Hello @leemannd sorry for the late reply. I'll create a PR to your branch then.

AaronHForgeFlow reviewed (commented):
Edited: dismissed comment

mdietrichc2c committed:
Add SO line description

Commit:
add Delivery Address to wizard. by default selec SO partner address.

leemannd committed:
Merge pull request #10 from leemannd/leemannd_add-sale_manual_delivery

add Delivery Address to wizard. by default selec SO partner address.

yvaucher commented:
2 errors in tests:

```
FAIL: test_00_sale_manual_delivery (odoo.addons.sale_manual_delivery.tests.test_manual_delivery.TestSaleStock)
` Test SO's manual delivery
Traceback (most recent call last):
`   File "/home/travis/build/OCA/sale-workflow/sale_manual_delivery/tests/test_manual_delivery.py", line 53, in test_00_sale_manual_delivery
`     should be created after "manual delivery" wizard call')
` AssertionError: stock.picking() is not true : Sale Manual Delivery: picking             should be created after "manual delivery" wizard call
FAIL: test_02_sale_various_manual_delivery (odoo.addons.sale_manual_delivery.tests.test_manual_delivery.TestSaleStock)
` Test SO's various manual delivery
Traceback (most recent call last):
`   File "/home/travis/build/OCA/sale-workflow/sale_manual_delivery/tests/test_manual_delivery.py", line 165, in test_02_sale_various_manual_delivery
`     should be created after "manual delivery" wizard call')
` AssertionError: stock.picking() is not true : Sale Manual Delivery: picking             should be created after "manual delivery" wizard call
FAILED
Module sale_manual_delivery: 2 failures, 0 errors
```

AaronHForgeFlow commented:
I just tested this module again and I think it works ok. Some minor issues found but can be easily fixed:

- Got error when trying to create the delivery for more than one SO. A warning message to the user telling that the operation is not allowed should be ok for me.

- If I cancel the SO and then I set it back to draft and then confirm the SO again. Then, when I create the deliveries those are not linked to the sale order, but they are actually created.

- The commented CI thing

@leemannd  I think it is worth it to move forward with this Pull Request. I will try to fix it if are good with that.

Regards.

leemannd commented:
@aheficent Thanks for the review and time taken. I had currently no time allowed to work on it :(
IMO we should close this in favor of your PR
